batcher/aligned-batcher/src/lib.rs

```diff
@@ -455,206 +341,183 @@ impl Batcher {
                         );
+                        std::mem::drop(batch_state_lock);
                         send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidNonce)
                             .await;
                         return Ok(());
                     }
-                    std::cmp::Ordering::Equal => {
-                        // if we are here nonce == expected_user_nonce
-                        if !self
-                            .handle_expected_nonce_message(
-                                batch_state,
-                                min_fee,
-                                nonced_verification_data,
-                                ws_conn_sink.clone(),
-                                client_msg.signature,
-                                addr,
-                            )
-                            .await
-                        {
-                            // message should not be added to batch
-                            return Ok(());
-                        };
-                    }
-                    std::cmp::Ordering::Greater => {
-                        // might be replacement message
-                        // if the message is already in the batch
-                        // we can check if we need to increment the fee
-                        // get the entry with the same sender and nonce
-                        if !self
-                            .handle_replacement_message(
-                                batch_state,
-                                nonced_verification_data,
-                                ws_conn_sink.clone(),
-                                client_msg.signature,
-                                addr,
-                                expected_user_nonce,
-                            )
-                            .await
-                        {
-                            // message should not be added to batch
-                            return Ok(());
-                        }
-                    }
-                }
+                };
+                let user_state = UserState::new(ethereum_user_nonce);
+                batch_state_lock.user_states.insert(addr, user_state);
+            }
+        }
 
-                info!("Verification data message handled");
+        // * ---------------------------------------------------*
+        // *        Perform validations over user state         *
+        // * ---------------------------------------------------*
+
+        // For now on until the message is fully processed, the batch state is locked
+        // This is needed because we need to query the user state to make validations and
+        // finally add the proof to the batch queue.
+
+        // At this point, we will have a user state for sure, since we have inserted it
+        // if not already present. It should be safe to call `unwrap()` here.
+        let batch_state_lock = self.batch_state.lock().await;
```

Comment:
**NOTE**: From here onwards, the batch state is locked. This means that the `batch_queue` **and** the `users_states` are both locked. I tried to decouple both but faced a lot of deadlocks. This might be a skill issue from my side, but didn't want to take so long on this PR and over complexify things. This could be explored and tackled in another PR that just focuses on that.
For the time being, this just replicates the behavior we had until this moment, since we were already locking all the batch state.

Reply:
This kind of comment is better placed in the code itself.